File: products-microservice/src/router/router.go
package router

import (
	"github.com/dieg0code/products-microservice/src/controllers"
	"github.com/dieg0code/products-microservice/src/db"
	"github.com/gin-gonic/gin"
)

type Router struct {
	ProductController controllers.ProductController
}

func NewRouter(productController controllers.ProductController) *Router {
	return &Router{
		ProductController: productController,
	}
}

func (r *Router) InitRoutes() *gin.Engine {
	gin.SetMode(gin.ReleaseMode)
	router := gin.New()
	router.Use(gin.Recovery())

	router.GET("", func(ctx *gin.Context) {
		ctx.JSON(200, gin.H{
			"message": "Welcome to Products Microservice",
		})
	})

	router.GET("/health", func(ctx *gin.Context) {
		ctx.JSON(200, gin.H{
			"message": "Products Microservice is healthy",
		})
	})

	router.GET("/ready", func(ctx *gin.Context) {

		dbConn := db.DatabaseConnection()
		err := db.CheckDatabaseConnection(dbConn)
		if err != nil {
			ctx.JSON(500, gin.H{
				"message": "Products Microservice is not ready",
			})
			return
		}

		ctx.JSON(200, gin.H{
			"message": "Products Microservice is ready",
		})
	})

	baseRoute := router.Group("/api/v1")
	{
		productRoute := baseRoute.Group("/products")
		{
			productRoute.POST("", r.ProductController.CreateProduct)
			productRoute.GET("/:productID", r.ProductController.GetProductById)
			productRoute.GET("", r.ProductController.GetAllProducts)
			productRoute.GET("/category/:category", r.ProductController.GetByCategory)
			productRoute.PUT("/:productID", r.ProductController.UpdateProduct)
			productRoute.DELETE("/:productID", r.ProductController.DeleteProduct)
		}
	}

	return router
}

File: sales-microservice/src/routes/router.ts
import { Router } from "express";
import { SaleController } from "../controllers/sale_controller";

export class AppRouter {
    private saleController: SaleController;

    constructor(saleController: SaleController) {
        this.saleController = saleController;
    }

    public initRoutes(): Router {
        const router = Router();

        router.get("/", (_req, res) => {
            res.json({ message: "Welcome to Sales Microservice" });
            
        });

        router.get("/health", (_req, res) => {
            res.json({ message: "Sales Microservice is healthy" });
        });

        const baseRoute = Router();

        const salesRoute = Router();

        salesRoute.get("/:saleID", this.saleController.getSaleByID.bind(this.saleController));
        salesRoute.get("/user/:userID", this.saleController.getSalesByUserID.bind(this.saleController));
        salesRoute.get("/date/:date", this.saleController.getSalesByDate.bind(this.saleController));
        salesRoute.get("/", this.saleController.getAllSales.bind(this.saleController));
        salesRoute.post("/", this.saleController.createSale.bind(this.saleController));
        salesRoute.delete("/:saleID", this.saleController.deleteSale.bind(this.saleController));

        baseRoute.use("/sales", salesRoute);

        router.use("/api/v1", baseRoute);

        return router;
        
    }
}